contracts/YieldFarm.sol

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.6.0;
+
+import "@openzeppelin/contracts-ethereum-package/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts-ethereum-package/contracts/utils/ReentrancyGuard.sol";
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "./interfaces/IStaking.sol";
+
+
+contract YieldFarm is ReentrancyGuardUpgradeSafe {
+
+    // lib
+    using SafeMath for uint;
+
+    // constants
+    uint constant TOTAL_DISTRIBUTED_AMOUNT = 800000;
+
+    uint NR_OF_EPOCHS = 24;
+
+    uint EPOCH_DURATION = 604800; // 7 days * 24 hours * 60 mins * 60 seconds
+
+    // structs
+
+    struct Epoch {
+        bool init;
+        uint stakeEpoch;
+        uint poolSize;
+        mapping (address => bool) claimed;
+    }
+
+
+    // state variables
+    IERC20 private _barn;
+    address private _usdc;
+    address private _susd;
+    address private _dai;
+    address private _barnYCurve;
+    IStaking private _staking;
+    mapping (uint => Epoch) epochs;
+
+
+    uint epochStart;
+
+
+    // modifiers
+
+
+
+
+    // construct
+
+    constructor(address barnBridgeAddress, address usdc, address susd, address dai, address barnYCurve, address stakeContract) public {
```

Comment:
barnBridgeAddress -> bondTokenAddress

Reply:
barnYCurve -> should not be part of the constructor -> should be calculated using the deterministic function most probably between _bond and _usdc UNI LP.

```/// https://uniswap.org/docs/v2/smart-contract-integration/getting-pair-addresses/
function genUniAddr(address left, address right) internal pure returns (address) {
    address first = left < right ? left : right;
    address second = left < right ? right : left;
    address pair = address(uint(keccak256(abi.encodePacked(
      hex'ff',
      uniFactory,
      keccak256(abi.encodePacked(first, second)),
      hex'96e8ac4277198ff8b6f785478aa9a39f403cb768dd02cbee326c3e7da348845f'
    ))));
    return pair;
}